File: common/hogvm/typescript/src/objects.ts
import { HogCallable, HogClosure, HogDate, HogDateTime, HogError, HogUpValue } from './types'

export function isHogDate(obj: any): obj is HogDate {
    return obj && typeof obj === 'object' && '__hogDate__' in obj && 'year' in obj && 'month' in obj && 'day' in obj
}

export function isHogDateTime(obj: any): obj is HogDateTime {
    return obj && typeof obj === 'object' && '__hogDateTime__' in obj && 'dt' in obj && 'zone' in obj
}

export function isHogError(obj: any): obj is HogError {
    return obj && typeof obj === 'object' && '__hogError__' in obj && 'type' in obj && 'message' in obj
}

export function newHogError(type: string, message: string, payload?: Record<string, any>): HogError {
    return {
        __hogError__: true,
        type: type || 'Error',
        message: message || 'An error occurred',
        payload,
    }
}

export function isHogCallable(obj: any): obj is HogCallable {
    return (
        obj &&
        typeof obj === 'object' &&
        '__hogCallable__' in obj &&
        'argCount' in obj &&
        'ip' in obj &&
        // 'chunk' in obj &&  // TODO: enable after this has been live for some hours
        'upvalueCount' in obj
    )
}

export function isHogClosure(obj: any): obj is HogClosure {
    return obj && typeof obj === 'object' && '__hogClosure__' in obj && 'callable' in obj && 'upvalues' in obj
}

export function newHogClosure(callable: HogCallable, upvalues?: number[]): HogClosure {
    return {
        __hogClosure__: true,
        callable,
        upvalues: upvalues ?? [],
    }
}

export function newHogCallable(
    type: HogCallable['__hogCallable__'],
    {
        name,
        chunk,
        argCount,
        upvalueCount,
        ip,
    }: {
        name: string
        chunk: string
        argCount: number
        upvalueCount: number
        ip: number
    }
): HogCallable {
    return {
        __hogCallable__: type,
        name,
        chunk: chunk,
        argCount,
        upvalueCount,
        ip,
    } satisfies HogCallable
}

export function isHogUpValue(obj: any): obj is HogUpValue {
    return (
        obj &&
        typeof obj === 'object' &&
        '__hogUpValue__' in obj &&
        'location' in obj &&
        'closed' in obj &&
        'value' in obj
    )
}

export function isHogAST(obj: any): boolean {
    return obj && ((typeof obj === 'object' && '__hx_ast' in obj) || (obj instanceof Map && obj.get('__hx_ast')))
}

File: common/hogvm/python/objects.py
from dataclasses import dataclass
from typing import Any, Optional


@dataclass
class CallFrame:
    ip: int
    chunk: str
    stack_start: int
    arg_len: int
    closure: dict


@dataclass
class ThrowFrame:
    call_stack_len: int
    stack_len: int
    catch_ip: int


def is_hog_date(obj: Any) -> bool:
    return isinstance(obj, dict) and "__hogDate__" in obj and "year" in obj and "month" in obj and "day" in obj


def is_hog_datetime(obj: Any) -> bool:
    return isinstance(obj, dict) and "__hogDateTime__" in obj and "dt" in obj and "zone" in obj


def is_hog_error(obj: Any) -> bool:
    return isinstance(obj, dict) and "__hogError__" in obj and "type" in obj and "message" in obj


def new_hog_error(type: str, message: Any, payload: Any = None) -> dict:
    return {
        "__hogError__": True,
        "type": type or "Error",
        "message": message or "An error occurred",
        "payload": payload,
    }


def is_hog_callable(obj: Any) -> bool:
    return (
        isinstance(obj, dict)
        and "__hogCallable__" in obj
        and "argCount" in obj
        and "ip" in obj
        # and "chunk" in obj # TODO: enable after this has been live for some hours
        and "upvalueCount" in obj
    )


def is_hog_closure(obj: Any) -> bool:
    return isinstance(obj, dict) and "__hogClosure__" in obj and "callable" in obj and "upvalues" in obj


def new_hog_closure(callable: dict, upvalues: Optional[list] = None) -> dict:
    return {
        "__hogClosure__": True,
        "callable": callable,
        "upvalues": upvalues or [],
    }


def new_hog_callable(type: str, arg_count: int, upvalue_count: int, ip: int, name: str, chunk: str) -> dict:
    return {
        "__hogCallable__": type,
        "name": name,
        "chunk": chunk,
        "argCount": arg_count,
        "upvalueCount": upvalue_count,
        "ip": ip,
    }


def is_hog_upvalue(obj: Any) -> bool:
    return (
        isinstance(obj, dict)
        and "__hogUpValue__" in obj
        and "location" in obj
        and "closed" in obj
        and "value" in obj
        and "id" in obj
    )


def is_hog_interval(obj: Any) -> bool:
    return isinstance(obj, dict) and obj.get("__hogInterval__") is True


def to_hog_interval(value: int, unit: str):
    return {
        "__hogInterval__": True,
        "value": value,
        "unit": unit,
    }
